content/post/asdf.md

```diff
@@ -0,0 +1,240 @@
+is that we can serialize literally any object, except for few rare types
+like mutexes or executable code. The implied drawbacks are substantial though:
+we should not load a pickle from unreliable source because it is vulnerable
+to remote code execution attacks; the loader must implement a virtual machine,
+which is hard with other programming languages; the memory consumption
+during serialization and deserialization can grow far beyond the original object size.
+The latter is important for machine learning and data science in particular:
+the objects can grow very big, say, bigger than a few hundred megabytes and
+pickling them is very slow and very demanding for the free operating memory.
+We witnessed 2 and even 3 times bigger memory consumption during
+`pickle`-ing which led to funny situtations when one has successfully computed
+the data but cannot dump it on disk so the whole effort becomes useless.
+
+So when we started Modelforge, we searched for the best serialization format
+which is **not** `pickle`.
+
+## Life beyond `pickle`
+
+Our best-shot requirements were:
+
+1. Binary format. It is impossible to efficiently save a huge dense tensor in JSON or YAML.
+2. At the same time, it would be nice to be able to save any JSON-like metadata without pain.
+3. The schema is optional. We did not want to force people to write the schema definitions - it is asking too much from ML researchers. Schema-ful formats separate it from the data and don't include
+into the file, and thus it becomes hard to introspect. Pythonistas take introspection for granted, to say at least.
+4. The typical data size may span over tens of gigabytes. Compared to gigabytes,
+the overhead of including a schema looks irrelevant indeed, hence the files
+become self-descriptive and the schema turns into a validation perk.
+This requirement rules out Protocol Buffers completely, because there can be
+performance and memory problems after 50MB already. It also excludes other
+popular binary serialization formats because they always target messaging and RPC.
+5. Python should be a first-class citizen. numpy arrays should be serializable
+without any additional code.
+6. Yet it should not be nailed to Python. The user-facing applications at source{d}
+are written in Go and otherwise it will be hard to integrate.
+7. On-the-fly compression. NLP models often contain strings which can require much
+space while being perfectly compressable. Integers can be compressed too since
+we sometimes don't know their range beforehand and use 32 bits while only 16 are
+really needed.
+
+[HDF5](https://support.hdfgroup.org/HDF5/) was the closest to those. It is binary,
+there is no schema, it supports big tensors, Python bindings are mature and
+well-integrated, there bindings for other languages. HDF5 is used in e.g. [Keras](https://github.com/keras-team/keras).
+However, it is not ideal in terms of performance and is also morally outdated.
```

Comment:
morally?!?

Reply:
Yeah, like CORBA or `ed`. Technically, they work, but you don't want to use them... Any better adjective?

Reply:
Rephrazed